internal/sourcemap/metadata.go

```diff
@@ -0,0 +1,254 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package sourcemap
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/go-sourcemap/sourcemap"
+	"github.com/pkg/errors"
+
+	"github.com/elastic/apm-server/internal/elasticsearch"
+	"github.com/elastic/apm-server/internal/logs"
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/elastic/go-elasticsearch/v8/esapi"
+)
+
+type MetadataCachingFetcher struct {
+	esClient         *elasticsearch.Client
+	set              map[Identifier]string
+	mu               sync.RWMutex
+	backend          Fetcher
+	logger           *logp.Logger
+	once             sync.Once
+	index            string
+	invalidationChan chan<- Identifier
+}
+
+func NewMetadataCachingFetcher(
+	c *elasticsearch.Client,
+	backend Fetcher,
+	index string,
+	invalidationChan chan<- Identifier,
+) *MetadataCachingFetcher {
+	return &MetadataCachingFetcher{
+		esClient:         c,
+		index:            index,
+		set:              make(map[Identifier]string),
+		backend:          backend,
+		logger:           logp.NewLogger(logs.Sourcemap),
+		invalidationChan: invalidationChan,
+	}
+}
+
+func (s *MetadataCachingFetcher) Fetch(ctx context.Context, name, version, path string) (*sourcemap.Consumer, error) {
+	if len(s.set) == 0 {
+		// If cache is empty and we're trying to fetch a sourcemap
+		// make sure the initial cache population has ended.
+		s.once.Do(func() {
```

Comment:
What happens if ES is not yet reachable at this point but later?

Reply:
Looking back at this, I don't think it's a good idea, this will basically block until the cache is populated since we cannot retrieve sourcemaps until then with the current implementation.

I'm not sure what's the best way to proceed here. Returning nothing might be misleading. Should we bypass the cache and send the request to ES if the cache is not populated yet ?

Reply:
This would be if the metadata cache is not ready. If a sourcemap is part of the metadata cache but the body is not yet  cached, it will be fetched from ES on request. 
So let's take a look at the metadata cache. What are the scenarios for when the cache might not be ready?
- ES is not reachable, due to a restart, misconfiguration, etc.: in this case, trying to fetch directly from ES would not succeed either. 
- Fetching metadata from ES is ongoing. The fetching and cache update for metadata can be expected to be fairly quick, even when many sourcemaps are stored. Bypassing the metadata cache would work here, but seems like a hack. I'd rather either block the request processing (and collect logs & metrics so we get an understanding how long this blocking period is) or treat it as _no sourcemap is available_ (and again, collect logs & metrics for the scenario). The third option would be to block the APM Server listeners until everything is readily set up, but this would definitely lead to apm event loss, which is definitely worse.

Reply:
> ES is not reachable, due to a restart, misconfiguration, etc.: in this case, trying to fetch directly from ES would not succeed either.

I don't think the metadata cache should be concerned If ES is not ready/reachable/available. A failure in that scenario would be "working as intended" imo because the metadata cache is proxying what it got from the ES fetcher (error, timeout, etc.).

> Fetching metadata from ES is ongoing. The fetching and cache update for metadata can be expected to be fairly quick, even when many sourcemaps are stored. Bypassing the metadata cache would work here, but seems like a hack. I'd rather either block the request processing (and collect logs & metrics so we get an understanding how long this blocking period is) or treat it as no sourcemap is available (and again, collect logs & metrics for the scenario). The third option would be to block the APM Server listeners until everything is readily set up, but this would definitely lead to apm event loss, which is definitely worse.

I was thinking of "if init is not completed, forward the request to the backend fetcher". In theory the first request for each sourcemap would be forwarded to ES anyway so this wouldn't create additional work unless we get a request for a missing sourcemap on ES during init. The reasoning behind this was that when the metadata cache is not initialized:
- if the sourcemap exists on ES: the metadata fetcher will send the request to the es fetcher anyway. No additional time/work/blocking: this would be the same as the metadata cache receiving a request for the first time. 
- if the sourcemap does not exist on ES: the metadata will send the request to the ES fetcher, creating additional work. The additional processing time is offset by the fact we would have had to wait for the init phase to finish if we had blocked so my assumption was that there wasn't gonna be a time difference between the two in the end.

Reply:
So this is really used for triggering the first fetch, in case it hasn't yet been initiated?

Reply:
The logic has been reworked and now we forward the request to ES if the metadata cache is not ready. 

This does not create additional costs since the response is cached in the lru cache.